ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,203 @@
+import           Data.Maybe (isNothing, fromJust)
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Ratio
+import           Data.Word
+import qualified Network.Socket as Socket
+import           System.Random
+
+import           Cardano.Slotting.Slot (SlotNo)
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAddress (..))
+import           Ouroboros.Network.Subscription.Ip
+import           Ouroboros.Network.Subscription.Dns
+
+import           Text.Printf
+
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+-- | Trace LedgerPeers events.
+data TraceLedgerPeers =
+      PickedPeer !RelayAddress !AckPoolStake ! PoolStake
+      -- ^ Trace for a peer picked with ackumulated and relative stake of its pool.
+    | PickedPeers !Word16 ![RelayAddress]
+      -- ^ Trace for the number of peers we wanted to pick and the list of peers picked.
+    | FetchingNewLedgerState !Int
+      -- ^ Trace for fetching a new list of peers from the ledger. Int is the number of peers
+      -- returned.
+
+
+instance Show TraceLedgerPeers where
+    show (PickedPeer addr ackStake stake) =
+        printf "PickedPeer %s ack stake %s ( %.04f) relative stake %s ( %.04f )"
+            (show addr)
+            (show $ unAckPoolStake ackStake)
+            (fromRational (unAckPoolStake ackStake) :: Double)
+            (show $ unPoolStake stake)
+            (fromRational (unPoolStake stake) :: Double)
+    show (PickedPeers n peers) =
+        printf "PickedPeers %d %s" n (show peers)
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of the stakepool. A value in the [0, 1] range.
+newtype PoolStake = PoolStake { unPoolStake :: Rational } deriving (Eq, Fractional, Num, Ord, Show)
+
+-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
+newtype AckPoolStake = AckPoolStake { unAckPoolStake :: Rational } deriving (Eq, Num, Ord)
+
+-- | Convert a list of pools with stake to a Map keyed on the ackumulated stake.
+--
+ackPoolStake :: [(PoolStake, NonEmpty RelayAddress)]
+             -> Map AckPoolStake (PoolStake, NonEmpty RelayAddress)
+ackPoolStake pl =
+    let pl' = reRelativeStake pl
+        ackList = snd $ foldl' fn (AckPoolStake 0, []) pl' in
+    Map.fromList ackList
+  where
+    fn :: (AckPoolStake, [(AckPoolStake, (PoolStake, NonEmpty RelayAddress))])
+       -> (PoolStake, NonEmpty RelayAddress)
+       -> (AckPoolStake, [(AckPoolStake, (PoolStake, NonEmpty RelayAddress))])
+    fn (as, ps) (s, rs) =
+        let ackst = AckPoolStake (unPoolStake s) in
```

Comment:
Nice, I like that it's now visible that we coerce `PoolStake` to `AckPoolStake`.